sklearn/feature_selection/_base.py

```diff
@@ -78,15 +79,22 @@ def transform(self, X):
         X_r : array of shape [n_samples, n_selected_features]
             The input samples with only the selected features.
         """
-        # note: we use _safe_tags instead of _get_tags because this is a
-        # public Mixin.
-        X = self._validate_data(
-            X,
-            dtype=None,
-            accept_sparse="csr",
-            force_all_finite=not _safe_tags(self, key="allow_nan"),
-            reset=False,
-        )
+        output_config_dense = _get_output_config("transform", estimator=self)["dense"]
+        if hasattr(X, "iloc") and output_config_dense == "pandas":
+            # Only check feature names and n_features when output is a dataframe
+            # This allow _transform to preserve `X`'s dtype
```

Comment:
Yes, it does work with `pass`, but we still want to check if the input is valid. For example, we enforce that the column names are consistent between `fit` and `transform` in all the transformers:

```python
from sklearn.feature_selection import SelectPercentile
import pandas as pd

X = pd.DataFrame({"a": [1, 10, 5], "b": [3, 1, 6], "c": [5, 6, 3], "d": [5, 1, 6]})
y = [1, 0, 1]

selector = SelectPercentile(percentile=50).set_output(transform="pandas")
selector.fit_transform(X, y)

X_test = X.copy()
X_test.columns = ["f1", "f2", "f3", "f4"]

# Errors because column names are not consistent.
selector.transform(X_test)
```

Reply:
It makes sense to do the validation. The thing that puzzled me is where/what magic was happening to make things work with respect to the PR. I like your updated comment.

Reply:
Ah, I see. What makes this PR work is the [update to `_transform`](https://github.com/thomasjpfan/scikit-learn/blob/17fb2d18093c8158f93bf9e103d380070e20ee69/sklearn/feature_selection/_base.py#L112). As long as `X` is a dataframe, `_safe_indexing(X, mask, axis=1)` will mask it correctly and return a dataframe.

Reply:
Exactly. This is what I thought, and then I spent a bunch of time trying to understand why the stuff here was required (the comment made me believe it was required). All good now.